File: spring-data-requery-test/src/main/java/org/springframework/data/requery/domain/upsert/AbstractUpsertTag.java
package org.springframework.data.requery.domain.upsert;

import io.requery.*;
import io.requery.query.MutableResult;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.requery.domain.AbstractPersistable;
import org.springframework.data.requery.domain.ToStringBuilder;

import java.util.UUID;

/**
 * @author Diego on 2018. 6. 10..
 */
@Getter
@Entity(name = "UpsertTag")
@Table(name = "upsert_tag")
public abstract class AbstractUpsertTag extends AbstractPersistable<UUID> {

    @Key
    protected UUID id;

    @Column
    protected String name;

    @ManyToMany(cascade = { CascadeAction.DELETE, CascadeAction.SAVE })
    protected MutableResult<AbstractUpsertEvent> events;

    @Override
    public int hashCode() {
        return name != null ? name.hashCode() : System.identityHashCode(this);
    }

    @Transient
    @Override
    protected @NotNull ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("name", name);
    }
}

File: spring-data-requery-kotlin/src/test/kotlin/com/coupang/springframework/data/requery/domain/upsert/AbstractUpsertTag.kt
package com.coupang.springframework.data.requery.domain.upsert

import com.coupang.kotlinx.objectx.ToStringBuilder
import com.coupang.springframework.data.requery.domain.AbstractPersistable
import io.requery.*
import io.requery.query.MutableResult
import java.util.*

/**
 * AbstractUpsertTag
 *
 * @author debop@coupang.com
 * @since 18. 5. 31
 */
@Entity
abstract class AbstractUpsertTag: AbstractPersistable<UUID>() {

    @get:Key
    abstract var id: UUID?

    @get:Column
    abstract var name: String?

    @get:ManyToMany(cascade = [CascadeAction.DELETE, CascadeAction.SAVE])
    abstract val events: MutableResult<AbstractUpsertEvent>


    override fun hashCode(): Int {
        return name?.hashCode() ?: System.identityHashCode(this)
    }

    @Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("name", name)
    }
}